src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultCollectionValidationStrategy.cs

```diff
@@ -60,9 +63,16 @@ private DefaultCollectionValidationStrategy()
             return new Enumerator(metadata.ElementMetadata, key, enumerator);
         }
 
-        public static IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
+        public IEnumerator GetEnumeratorForElementType(ModelMetadata metadata, object model)
         {
-            var getEnumeratorMethod = _getEnumerator.MakeGenericMethod(metadata.ElementType);
+            MethodInfo getEnumeratorMethod;
+            if (!_genericGetEnumeratorCache.TryGetValue(metadata.ElementType, out getEnumeratorMethod))
+            {
+                getEnumeratorMethod = _genericGetEnumeratorCache.AddOrUpdate(
+                        key: metadata.ElementType,
+                        addValueFactory: (type) => _getEnumerator.MakeGenericMethod(type),
+                        updateValueFactory: (type, methodInfo) => methodInfo);
+            }
             return (IEnumerator)getEnumeratorMethod.Invoke(null, new object[] { model });
```

Comment:
We might want to think more carefully about the issue that this code solves and see if there's a better way.

This is for cases where a type implements multiple kinds of `IEnumerator<T>`. We've already decided what we think `T` is, but there can be problems calling the non-generic `GetEnumerator` because we don't know which enumerator it's going to return.... Maybe we want to no-op if the results don't match our expectations....

We hit this case with `JObject`.